### ISSUE TITLE
opal-parser of Opal 1.0 fails to parse non-ascii comments

### ISSUE THREAD

**Issue by yhara** [opal/opal#2066]
Given this test.rb:

```rb
require 'opal-parser'
eval("# こんにちは
p 1")
```

Opal 0.11.4 works fine.

```
% opal _0.11.4_ test.rb
Object freezing is not supported by Opal
1
```

Opal >= 1.0 raises the following error.

```
% opal _1.0.0_ test.rb
Failed to load WithCLexer, using pure Ruby lexer
Object freezing is not supported by Opal

/private/var/folders/01/8_5tmzpj66z4fp1y27j2q9dw0000gn/T/opal-nodejs-runner-20200118-48507-o2uodj:4665
      throw exception;
      ^
(eval):in `p 1'
encode: undefined method `encode' for nil
    at constructor.$$method_missing (/private/var/folders/01/8_5tmzpj66z4fp1y27j2q9dw0000gn/T/opal-nodejs-runner-20200118-48507-o2uodj:3907:56)
    at constructor.method_missing_stub (/private/var/folders/01/8_5tmzpj66z4fp1y27j2q9dw0000gn/T/opal-nodejs-runner-20200118-48507-o2uodj:1310:35)
    at constructor.$$slice (/private/var/folders/01/8_5tmzpj66z4fp1y27j2q9dw0000gn/T/opal-nodejs-runner-20200118-48507-o2uodj:26680:52)
    at constructor.$$tok (/private/var/folders/01/8_5tmzpj66z4fp1y27j2q9dw0000gn/T/opal-nodejs-runner-20200118-48507-o2uodj:38128:33)
    at constructor.$$advance (/private/var/folders/01/8_5tmzpj66z4fp1y27j2q9dw0000gn/T/opal-nodejs-runner-20200118-48507-o2uodj:30171:30)
    at constructor.$$next_token (/private/var/folders/01/8_5tmzpj66z4fp1y27j2q9dw0000gn/T/opal-nodejs-runner-20200118-48507-o2uodj:40836:28)
    at $$7 (/private/var/folders/01/8_5tmzpj66z4fp1y27j2q9dw0000gn/T/opal-nodejs-runner-20200118-48507-o2uodj:25876:29)
    at Object.Opal.yieldX (/private/var/folders/01/8_5tmzpj66z4fp1y27j2q9dw0000gn/T/opal-nodejs-runner-20200118-48507-o2uodj:1475:18)
    at constructor.$Kernel_catch$64 (/private/var/folders/01/8_5tmzpj66z4fp1y27j2q9dw0000gn/T/opal-nodejs-runner-20200118-48507-o2uodj:4845:21)
    at Opal.send (/private/var/folders/01/8_5tmzpj66z4fp1y27j2q9dw0000gn/T/opal-nodejs-runner-20200118-48507-o2uodj:1671:19)
```

**Comment by elia** [comment]
@iliabylich you have any clue about where to look for where those ranges are produced?
you can see below it's producing ranges that probably count the bytes instead of the graphemes (or whatever they're called).

master:

```
opal:elia/non-ascii-comments ⤑ bin/opal tmp/non-ascii-comments.rb                                                                                      ~/C/opal
parser/source/buffer.rb:199:16: warning: `*' interpreted as argument prefix
parser/source/buffer.rb:199:           puts *caller if $DEBUG
parser/source/buffer.rb:199:                ^                
parser/source/range.rb:38:14: warning: `*' interpreted as argument prefix
parser/source/range.rb:38:         puts *caller if $DEBUG
parser/source/range.rb:38:              ^                
Object freezing is not supported by Opal
parser/source/range.rb:38
corelib/runtime.js:1675
corelib/class.rb:39
parser/lexer.rb:23385
parser/lexer.rb:23424
parser/lexer.rb:19444
parser/base.rb:231
racc/parser.rb:293
corelib/runtime.js:1479
#<Parser::Source::Range (eval) 0...17>
"********************************************************************************"
["# こんにちは\np 1", 0...17, "# こんにちは\np 1", "# こんにちは\np 1", #<Encoding:UTF-16LE>]
parser/source/buffer.rb:199
parser/source/range.rb:135
parser/source/comment.rb:56
corelib/runtime.js:1675
corelib/class.rb:39
parser/lexer.rb:23424
parser/lexer.rb:19444
parser/base.rb:231
racc/parser.rb:293
"********************************************************************************"
["# こんにちは\np 1", 0...17, "# こんにちは\np 1", "# こんにちは\np 1", #<Encoding:UTF-16LE>]
parser/source/buffer.rb:199
parser/lexer.rb:23381
parser/lexer.rb:23428
parser/lexer.rb:19444
parser/base.rb:231
racc/parser.rb:293
corelib/runtime.js:1479
corelib/kernel.rb:676
corelib/runtime.js:1675
parser/source/range.rb:38
corelib/runtime.js:1675
corelib/class.rb:39
parser/lexer.rb:23385
parser/lexer.rb:23428
parser/lexer.rb:19444
parser/base.rb:231
racc/parser.rb:293
corelib/runtime.js:1479
"********************************************************************************"
["# こんにちは\np 1", 18...19, nil, "# こんにちは\np 1", #<Encoding:UTF-16LE>]
parser/source/buffer.rb:199
parser/lexer.rb:23381
parser/lexer.rb:11043
parser/base.rb:231
racc/parser.rb:293
corelib/runtime.js:1479
corelib/kernel.rb:676
corelib/runtime.js:1675
racc/parser.rb:288
(eval):in `p 1'
encode: undefined method `encode' for nil
  from corelib/kernel.rb:3:11:in `new'
  from corelib/runtime.js:1314:7:in `method_missing_stub'
  from parser/source/buffer.rb:200:11:in `encode'
  from parser/lexer.rb:23381:5:in `slice'
  from parser/lexer.rb:11043:14:in `tok'
  from parser/base.rb:231:15:in `advance'
  from racc/parser.rb:293:34:in `next_token'
  from corelib/runtime.js:1479:5:in `Opal.yieldX'
  from corelib/kernel.rb:676:5:in `$Kernel_catch$66'
  from corelib/runtime.js:1675:7:in `Opal.send'
```

0.11.4

```
opal:elia/non-ascii-comments ⤑ cd ..; opal _0.11.4_ opal/tmp/non-ascii-comments.rb; cd -                                                               ~/C/opal
Object freezing is not supported by Opal
"********************************************************************************"
["# こんにちは\np 1", 0...7, "# こんにちは", "# こんにちは\np 1", #<Encoding:UTF-16LE>]
"********************************************************************************"
["# こんにちは\np 1", 0...7, "# こんにちは", "# こんにちは\np 1", #<Encoding:UTF-16LE>]
"********************************************************************************"
["# こんにちは\np 1", 8...9, "p", "# こんにちは\np 1", #<Encoding:UTF-16LE>]
"********************************************************************************"
["# こんにちは\np 1", 8...9, "p", "# こんにちは\np 1", #<Encoding:UTF-16LE>]
"********************************************************************************"
["# こんにちは\np 1", 10...11, "1", "# こんにちは\np 1", #<Encoding:UTF-16LE>]
1
Exited with status 0
opal:elia/non-ascii-comments ⤑                                                                                                                         ~/C/opal
```

tmp/non-ascii-comments.rb
```rb
require 'opal-parser'
$DEBUG = 1
eval("# こんにちは
p 1")
```

modified parser in buffer.rb#slice

```rb
      def slice(range)
        if @slice_source.nil?
          @source[range]
        else
          p '*'*80 if $DEBUG
          p [@slice_source, range, @slice_source[range], @source, @source.encoding] if $DEBUG
          @slice_source[range].encode(@source.encoding)
        end
      end
```

**Comment by elia** [comment]
Apparently patching the Lexer to forcefully use `U` to unpack like it was UTF-8 fixes the issue:

```rb
  class Parser::Lexer
    def source_buffer=(source_buffer)
      @source_buffer = source_buffer

      if @source_buffer
        source = @source_buffer.source
        # Force UTF8 unpacking even if JS works with UTF-16/UCS-2
        # See: https://mathiasbynens.be/notes/javascript-encoding
        @source_pts = source.unpack('U*')
      else
        @source_pts = nil
      end
    end
  end
```

_PR incoming_

**Comment by elia** [comment]
Fixed by #2074
